Adding tuple_cast to the casts submodule

jonmooser opened the pull request:
This should be a useful addition, hopefully covering a reasonable set of use cases.  I added tests and documentation.

One open question might be returning collection types other that list. For example, would tuple_cast be useful?

Maybe not needed since 
key(cast=tuple(list_cast(...) ) )
is easy enough.

jonmooser committed:
Adding list_cast to the casts module

This takes string representation of a list and returns python list.
The default behaviour to use a comma as the delimiter, but alternate delimiters can be specified as well.

If an optional base_cast function is supplied, it will be called on each member of the list.

A set of tests are added to test_casts.py to test all of the params.

jonmooser committed:
Updating version and changelog

jonmooser committed:
Running black formatter

jonmooser committed:
Fixing mypy errors

jonmooser committed:
Last patch introduced a test coverage gap... fixing

bwindsor commented on typedconfig/casts.py:
```diff
@@ -1,14 +1,111 @@
+
+@overload
+def list_cast(
+    base_cast: Optional[Callable[[str], T]] = ...,
```
I think this line should just be
```python
base_cast: Callable[[str, T]],
```

I'm not sure I'm 100% clear on the nuances of `@overload`, but as far as I understand, these overloads are trying to deal with the two cases:
(a) User does not provide a value for `base_cast` or provides a value for `base_cast` of `None`. In this case we will be casting to `List[str]`
(b) User provides a function to `base_cast`. In this case we will be casting to `List[T]`

In case (b), therefore, a value is provided for `base_cast` and therefore the overload should not include the `= ...` which indicates that the input is optional. Similarly the `Optional` is not required since in this case the user is not providing a value of `None` since that is handled by case (a).

Without this change, the type inference works correctly in mypy but PyCharm's autocomplete does not pick it up. With this change, both work.

jonmooser replied:
Good catch on the @overload. I'll fix that. 

And I'll change the whole thing to be tuple_cast. I did think of that - it's a bit odd to have a mutable type inside an immutable structure. There could be a use for other immutable collection types, like a frozen_set_cast, but I'd stick to just tuple_cast for now.

bwindsor reviewed (commented):
Thanks! Just a commend about one of the `@overload`s.
Regarding `key(cast=tuple(list_cast(...) ) )` this wouldn't work, since you'd actually need something like
```
key(cast=lambda x: tuple(list_cast(...)(x))
```
So probably a separate `tuple_cast` function would be useful which reuses the logic from `list_cast` would be good! To be fair, I can't really see much use for `list_cast` and perhaps we should always be using `tuple_cast`, since config values shouldn't be mutable once read in anyway!

jonmooser committed:
Change list_cast to tuple_cast

We agree that you should always want a tuple and never a list in this case.

Also fixed the overloads. Tested the Pylance type chacker and it seems to resolve to the right type
with all permutations of the args

jonmooser committed:
Fixing tyckcheck for python 3.7, 3.8

I installed python 3.7 locally and checked that mypy and pytest returned with no errors

bwindsor commented on typedconfig/casts.py:
```diff
@@ -1,14 +1,111 @@
+
+        # remove the final element if it is empty and we should ignore trailing delimiters
+        if ignore_trailing_delimiter and len(str_list[-1]) == 0:
+            str_list = str_list[:-1]
```
I think you can just use `str_list.pop()` here to save creating a whole new list when actually you just want to remove the last element from the existing one.

bwindsor commented on typedconfig/casts.py:
```diff
@@ -1,14 +1,111 @@
+        if base_cast is None:
+            return tuple(str_list)
+
+        return tuple([base_cast(s) for s in str_list])
```
Could just use `tuple(base_cast(s) for s in str_list)` to create the tuple directly from a generator expression, which saves creating the intermediate list.

bwindsor reviewed (commented):
A couple of optimisation things (probably don't make any difference realistically, but nice to do them since they're easy changes). Otherwise looks good!

jonmooser committed:
A couple of small code improvements
* Don't create intermediate list before creating tuple
* Use pop() to remove the last element from list

bwindsor commented:
Looks great, thanks @jonmooser!
